tasks/migrate-certificates.ts

```diff
@@ -30,12 +33,310 @@ interface MigratedCertificate {
+  inputData,
+  certificateContract,
+}: {
+  outputData: MigratedCertificates;
+  inputData: InputData[];
+  certificateContract: Certificate;
+}): Promise<Summary> => {
+  const totalCertificateSupplyOnChain = await outputData.reduce(
+    async (summary, certificate) => {
+      const updatedSummary = await summary;
+      const certificateIdAmount = await certificateContract.getPurchaseAmount(
+        certificate.tokenId!
+      );
+      updatedSummary.certificates.push(certificateIdAmount);
+      updatedSummary.sum = updatedSummary.sum.add(certificateIdAmount);
+      return updatedSummary;
+    },
+    Promise.resolve({ sum: Zero, certificates: [] as BigNumber[] })
+  );
+  const expectedTotalCertificateSupply = inputData
+    .map((d) => d.amounts)
+    .reduce((total, next) => {
+      return total + next.reduce((t, n) => t + n, 0);
+    }, 0);
+  return {
+    totalCertificateSupplyOnChain,
+    expectedTotalCertificateSupply,
+  };
+};
+
+const validateMigrateEvent = ({
+  txResult,
+  removalContract,
+  hre,
+  certificateIndex,
+  inputData,
+  recipient,
+}: {
+  txResult: ContractReceipt;
+  removalContract: Removal;
+  hre: CustomHardHatRuntimeEnvironment;
+  certificateIndex: number;
+  inputData: InputData[];
+  recipient: string;
+}): void => {
+  const eventLog: {
+    certificateId: number;
+    removalAmounts: number[];
+    removalIds: number[];
+    certificateAmount: number;
+    certificateRecipient: string;
+  } = parseTransactionLogs({
+    contractInstance: removalContract,
+    txReceipt: txResult,
+  })
+    .filter((log) => log.name === 'Migrate')
+    .flatMap((log) => ({
+      certificateId: log.args.certificateId.toNumber(),
+      removalAmounts: log.args.removalAmounts.map((a: BigNumber) =>
+        Number(hre.ethers.utils.formatUnits(a.mul(1_000_000), 18))
+      ),
+      removalIds: log.args.removalIds,
+      certificateAmount: Number(
+        hre.ethers.utils.formatUnits(
+          log.args.certificateAmount.mul(1_000_000),
+          18
+        )
+      ),
+      certificateRecipient: log.args.certificateRecipient,
+    }))[0];
+  const datastoreCertificate = inputData[certificateIndex];
+  const offChainAmountsMatchOnchainAmounts =
+    JSON.stringify(eventLog.removalAmounts) ===
+    JSON.stringify(datastoreCertificate.amounts);
+  const offChainIdsMatchOnchainIds =
+    JSON.stringify(eventLog.removalAmounts) ===
+    JSON.stringify(datastoreCertificate.amounts);
+  if (!offChainAmountsMatchOnchainAmounts) {
+    throw new Error(
+      `Removal amounts do not match for certificate ${eventLog.certificateId}. Expected: ${datastoreCertificate.amounts} , Got: ${eventLog.removalAmounts}`
+    );
+  }
+  if (!offChainIdsMatchOnchainIds) {
+    throw new Error(
+      `Removal ids do not match for certificate ${eventLog.certificateId}. Expected: ${datastoreCertificate.ids} , Got: ${eventLog.removalIds}`
+    );
+  }
+  if (eventLog.certificateAmount !== datastoreCertificate.data.gramsOfNrts) {
```

Comment:
wouldn't the certificateAmount be represented here as tonnes of carbon, in wei, which shouldn't equal `gramsOfNrts` ?

Reply:
on-chain values are represented as tonnes in wei (5 tonnes === 5 10**18). Above this ([here](https://github.com/nori-dot-eco/contracts/blob/jaycen-migration-validation-and-summary/tasks/migrate-certificates.ts#L127)) line we are formatting on-chain values as grams  so that these two are the same